### ISSUE TITLE
Combination of several stream operations leads to OOM   

### ISSUE THREAD

**Issue by sebver** [zio/zio#6399]
Running the following code snippet with scala-cli will almost immediately lead to an OOM. If you remove the stream concatenation (and just keep the `repeatEffect`), there are no issues. If you use `.grouped(...)` instead of the `chunkN` + `mapChunks` combination, it also works fine.

```
//> using scala "2"
//> using lib "dev.zio::zio:1.0.13"
//> using lib "dev.zio::zio-streams:1.0.13"
//> using `java-opt` "-Xmx400M"

import zio._
import zio.stream._

object Main extends zio.App {

  val bytes =
    ZStream.succeed[Byte](0x1) ++
      ZStream.repeatEffect(random.nextBytes(1024)).flattenChunks

  def run(args: List[String]): URIO[ZEnv, ExitCode] =
    bytes
      .chunkN(1024 * 1024)
      .mapChunks(Chunk.single)
      .runDrain
      .exitCode
}
```

This is coming from a real world situation: [zio-s3 does this in its `multipartUpload` streaming code](https://github.com/zio/zio-s3/blob/master/src/main/scala/zio/s3/Live.scala#L181). 

I was generating and writing a huge xml file, so `ZStream.succeed("<xml>") ++ content ++ ZStream.succeed("</xml>")` leads to accumulating everything in memory.

**Comment by vigoo** [comment]
Note: the new streams implementation also fails with OOM but much later

**Comment by luis3m** [comment]
```scala
import zio._
import zio.stream._

object Main extends zio.App {

  val bytes =
    ZStream.succeed[Byte](0x1) ++
      ZStream.repeatEffect(random.nextBytes(1024))

  def run(args: List[String]): URIO[ZEnv, ExitCode] =
    bytes
      .chunkN(1024 * 1024)
      .runDrain
      .exitCode
}
```
This slightly shorter version seems to fail as well.

**Comment by mingyang91** [comment]
I have the same problem. In my case, Use `intersperse()` and it will cause OOM also.

**Comment by mingyang91** [comment]
I've found the point of memory leak; and will post a new issue to discuss this later.

**Comment by mingyang91** [comment]
@sebver This has been fixed in #6807, I think it can be closed.
